### PR TITLE
Tdc_gamma

### PR DESCRIPTION
**Pull request by adamjermyn**
This PR implements:
1.  An improved TDC solver which always picks the slowest-decaying solution when there are multiple solutions present. This both improves the stability of the solution (avoiding jumping between solutions when parameters are perturbed) and gives smoother transitions from convective to radiative.
2. A change to the TDC equations, which results in TDC reducing to MLT Cox (within floating precision) in steady state, even in the inefficient limit.

The latest test suite run isn't complete, but I believe the only case with problems will be `ccsn_IIp`, and that the two `astero` cases will fail because they need recalibrating.

### CODE REVIEW

**Comment by warrickball** [comment]
> and that the two astero cases will fail because they need recalibrating.

`simplex_solar_calibration` and `example_astero` are now [passing](https://testhub.mesastar.org/tdc_gamma/commits/7f08197) again after rolling back to the previous, Cox-calibrated values of MLT α.

File: star/defaults/controls.defaults [hunk 1]
```diff
@@ -2310,7 +2310,7 @@
       ! Options are:
 
       ! + 'none'       : just give radiative values with no mixing.
-      ! + 'TDC'        : Time-dependent convection based on the Khufuss 1986 model. Reduces to Cox at long times.
+      ! + 'TDC'        : Time-dependent convection based on the Khufuss 1986 model. Reduces to Cox at long times and large optical depths.
```
[the remark this answers is not in the document]
Comment (adamjermyn) [reply star/defaults/controls.defaults:2313]: It now does reduce, sorry I forgot to change this line.

[the remark this answers is not in the document]
Comment (adamjermyn) [reply star/defaults/controls.defaults:2313]: Fixed!

**Comment by adamjermyn** [comment]
Awesome, that's a great way to confirm that the reduction indeed works :-).

Looks like it's just `ccsn_IIp` that's failing.

**Comment by orlox** [comment]
Digging into ccsn_IIp I've managed to get it past inlist_shock_part1 with the attached modifications (was working on `e2235b49`). Only change I did outside of the test case was in solver_support. Found the code was getting stuck on the fe56 equation so just scaled down the residuals of species artificially to push it forward (so, nasty workaround, not a solution). Still, this should be similar to turning off gold tolerances for the species, so I think what comes out is not total nonsense. I also made the tests with `max_abs_du_div_cs_for_convection=1d99`, in order to deactivate this option that shuts off convection and see if TDC could just take care of it. Keeping `max_abs_du_div_cs_for_convection` at its default I could not progress much.

The rest are mostly modifications to inlists that should help things (higher resolution, scaling max corrections). However I did have to shut off convection near the surface to prevent the outermost cells from going wild early on using `max_abs_du_div_cs_for_convection=0.9999d0`. I use a similar workaround often with `ppisn`, never really traced well the source of the issue. With these changes the run seems to die in `inlist_shock_part3`. Apparently a few of the surface cells still go wild in `inlist_shock_part2` and make it think the shock has reached the surface.

Nevertheless, one thing that comes out is that the star becomes fully convective by this point, with the inner 2 solar masses having convective velocities of a few times the sound of speed. The convective core that is formed after energy injection seems to propagate outwards faster than the shock, which makes me think that it is the result of highly supersonic convection that is not limited in TDC.

Unfortunately had an emergency trip come up and will be travelling till late Wednesday (might miss next mesa-dev also). Can get back to this on Thursday but just wanted to share my unsuccessful attempts so far in case someone else has an idea.

[changes.zip](https://github.com/MESAHub/mesa/files/8056669/changes.zip)

**Comment by adamjermyn** [comment]
Thanks for the update! This seems like progress!

Would it help if we added the option to limit the convection speed to the sound speed in TDC? I'd have to think a bit about how to do it, but my immediate thought is that we could just cap the output convection speed, then recalculate `Y_face` using the post-cap speed. That'd be pretty easy to do, so just let me know if that'd be useful.

**Comment by orlox** [comment]
That sounds like a reasonable approach, pretty much the same logic of the old option to limit convective velocities.

**Comment by orlox** [comment]
Another option which I use in the implicit rotation code is to not have a hard cap, which can be bad for convergence, but rather something that continuously asymptotes to a threashold. I use a sigmoid for this, defined by two parameters:
`f(x,A,B) = 2*(A-B)/(1+exp(-2*(x-B)/(A-B)))-A+2*B`
This function is such that it has a derivative of unity at x=B, and asymptotes to A when X goes to infinity. So then one can use something like
```fortran
if (conv_vel_div_cs > A) then
   conv_vel =cs*f(conv_vel_div_cs,A,B)
end if
```
this requires defining two parameters rather than just a cap, but will make the result of TDC continuous, and also have continuous derivatives.

**Comment by adamjermyn** [comment]
For now I've just added a hard limiter to TDC. Let's see how that works?
It's in a new branch (tdc_gamma_no_cs). Let me know if that helps…

-Adam

**Comment by warrickball** [comment]
Looks like `ccsn_IIp` still fails: :frowning_face: 

https://testhub.mesastar.org/tdc_gamma_no_cs/commits/b43a65b/test_cases/star/ccsn_IIp

**Comment by orlox** [comment]
`ccsn_IIp` passes with convection off. Also tried using a smooth limiter for convective velocities and that makes `5M_cepheid_blue_loop` pass (but who know what it could break). I just did that quickly based on Adam's changes in the attached turb.f90, making the limiter softly turn on from 0.9*csound upwards:

```fortran
      ! Cap at cs
      csound = convert(sqrt(P/rho)) ! Approximate
      A = 1d0*unconvert(csound)
      B = 0.9d0*unconvert(csound)
      if (conv_vel%val > B%val) then
         conv_vel = 2*(A-B)/(1+exp(-2*(conv_vel-B)/(A-B)))-A+2*B
         ! L = L0 * (gradL + Y) + c0 * Af * Y_env
         ! L = L0 * (gradL + Y) + c0 * sqrt_2_div_3 * csound * (Gamma / (1 + Gamma)) * Y
         ! L - L0 * gradL = Y * (L0 + c0 * sqrt_2_div_3 * csound * (Gamma / (1 + Gamma)))
         Y_face = unconvert(info%L - info%L0 * info%gradL) / (unconvert(info%L0) + unconvert(info%c0) * sqrt_2_div_3 * unconvert(csound) * (info%Gamma / (1d0 + 
      end if
```

where A and B are `auto_diff_real_star_order1`. But one can get it to run by simply removing this limiter. In any case I do think limiting convective velocities to the sound speed is a sound thing to do.

**Comment by adamjermyn** [comment]
Awesome! Can you commit the changes to `ccsn_IIp`? Then we can merge this in...

For limiting: I'd like to get Radek's take (because TDC is built for situations with shocks and such) just to make sure we're not double-counting... so maybe we can open an issue for that discussion? (and keep the no_cs branch open).

**Pull request merged**